File: python/otp_enc_d.py
#!/usr/bin/env python3
import signal
import socket
import sys
import tqdm

MAX_CONNECTION_REQ = 5
HEADER_SIZE = 10
BUFFER_SIZE = 8

class OTPEncodeServer():

    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.server_socket = None
        self.conn_socket = None
    
    def start_server(self):
        try: 
            #addres family: AF_INET(IPv4) 
            #socket type: SOCK_STREAM (TCP)
            self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM) #create new socket
            self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1) #makes address reusable
            print("[+]Server socket created")
        except OSError as e:
            print(f"[+]Unable to create socket: {e}")
            self.server_socket = None
        else:        
            try:
                self.server_socket.bind((self.host, self.port)) #bind socket to address
                print("[+]Binding server socket")
                self.server_socket.listen(MAX_CONNECTION_REQ)  #establishes listening socket - limited to 5 connect requests
                print("[+]Listening for incoming connections")
            except OSError as e:
                print(f"[+]Unable to bind socket: {e}")
                self.server_socket = None
        
        if self.server_socket is None:
            print("[+]Unable to establish server socket")
            sys.exit(1)

    def recv_data(self):
        full_data = ''
        new_data = True

        while True:
            data = self.conn_socket.recv(BUFFER_SIZE)
            if new_data:
                #print(f"[+]New message length: {data[:HEADER_SIZE]}")
                data_len = int(data[:HEADER_SIZE])
                new_data = False

            full_data += data.decode("utf-8")
            if len(full_data)-HEADER_SIZE == data_len:
                #print(f"[+]Message received: {full_data[HEADER_SIZE:]}")
                full_data  = full_data[HEADER_SIZE:]
                break

        return full_data
    
    def send_data(self, text):
        data = f'{len(text):<{HEADER_SIZE}}' + text
        #print(f"[+]Data sent: {data}")
        self.conn_socket.send(bytes(data, "utf-8"))

    def encode_text(self, plaintext, key):
        enc_key = {
            'A': 0, 'B': 1,  'C': 2,  'D': 3,  'E': 4,  'F': 5,  'G': 6,  'H': 7,  'I': 8,
            'J': 9, 'K': 10, 'L': 11, 'M': 12, 'N': 13, 'O': 14, 'P': 15, 'Q': 16, 'R': 17,
            'S': 18,'T': 19, 'U': 20, 'V': 21, 'W': 22, 'X': 23, 'Y': 24, 'Z': 25, ' ': 26, 
        }

        #convert each char into their numerical value
        ptext = [enc_key[letter] for letter in plaintext]
        ktext = [enc_key[letter] for letter in key]
        #take the sum of the plaintext char and the key char, modula 27, to get the value of the ciphertext char
        ctext = [ (p+k)%len(enc_key) for p,k in zip(ptext, ktext) ]
        #print(f"[+]Cipher values: {ctext}")

        #convert numerical representation to string
        ciphertext = []
        for val in ctext:
            #appends the key that matches the numerical value (val)
            ciphertext.append(list(enc_key.keys())[list(enc_key.values()).index(val)])
        ciphertext = ''.join(ciphertext) #converts list of chars to string
        #print(f"[+]Ciphertext: {ciphertext}")

        return ciphertext

    def handle_connection(self):
        connected = True
        while connected:
            self.conn_socket, clientaddr = self.server_socket.accept() #accept incoming connection
            print(f"[+]Connected to {clientaddr[0]}: {clientaddr[1]}")

            plaintext = self.recv_data() #receives plaintext message from client
            #print(f"[+]Plaintext: {plaintext}")

            self.send_data(str(len(plaintext))) #sends size of plaintext to client as a confirmation of receipt

            key = self.recv_data() #receives key from client
            #print(f"[+]Key: {key}")

            if len(plaintext) == len(key): #verify size of plaintext and key are equal
                print("[+]Begin encoding...")
                ciphertext = self.encode_text(plaintext, key)
                self.send_data(ciphertext) #send ciphertext to client
                print("[+]Encrypted message sent")
            else:
                print("[+]Unable to encode. Closing connection")
            self.conn_socket.close()    
            connected = False
    
    def sigint_handler(self, sig, frame):
        print("\n[+]Shutting down server")
        #self.server_socket.shutdown(socket.SHUT_RDWR)
        self.server_socket.close()
        print("[+]Goodbye")
        sys.exit(0)

if __name__ == "__main__":
    HOST = socket.gethostname()
    PORT = 5757  
    
    s = OTPEncodeServer(HOST, PORT)
    s.start_server()
    
    signal.signal(signal.SIGINT, s.sigint_handler)
    print("[+]Press Ctrl+C to shutdown server")
    
    while True:
        s.handle_connection()

File: c/otp_enc_d.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h> 
#include <sys/socket.h>
#include <netinet/in.h>
#include <signal.h>
#include <sys/wait.h>
#include "network_helpers.h"

#define BACKLOG 5
#define MAX_PIDS 5

static void altSIGTERM(int signo);
void CleanProcesses(int* child_pid_cnt);
//char* EncodeText(char* plaintext, char* keytext);

int child_pid_cnt = 0;

int main(int argc, char *argv[]){ 

    int port_number, socket_fd, connection_fd;
    socklen_t sz_client_info;
    struct sockaddr_in server_addr, client_addr;
    pid_t child_pid = -5;
    //char* msg_buffer;
    uint32_t ptext_size;
    char* ptext;
    uint32_t key_size;
    char* ktext;
    //char* enc_text;

    struct sigaction SIGTERM_action = {0};
    SIGTERM_action.sa_handler = altSIGTERM;
    sigfillset(&SIGTERM_action.sa_mask);
    SIGTERM_action.sa_flags = SA_RESTART;
    sigaction(SIGTSTP, &SIGTERM_action, NULL);

    if(argc < 2){
        fprintf(stderr, "otp_enc_d: Invalid number of arguments\n");
        exit(1);
    }

    port_number = atoi(argv[1]); //convert user entered port number to int

    if(port_number == 0){
        fprintf(stderr, "otp_enc_d: Invalid port number entered\n");
        exit(1);
    }    
    //set up address struct for server
    memset((char*)&server_addr, '\0', sizeof(server_addr)); //clear out memory of struct
    server_addr.sin_family = AF_INET; //create network capable socket
    server_addr.sin_port = htons(port_number); //htons() converts int from host byte order to network byte order
    server_addr.sin_addr.s_addr = INADDR_ANY; //allows any address for connection to server

    //set up socket
    socket_fd = socket(AF_INET, SOCK_STREAM, 0); //domain: IPV4; type: TCP; protocol: default  
    if(socket_fd < 0){
        perror("otp_enc_d listening socket set up error: ");
        exit(1);
    }
    //bind socket to address
    if(bind(socket_fd, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0){
        perror("otp_enc_d error binding socket: ");
        exit(1);
    }
    //begin listening on socket
    if(listen(socket_fd, BACKLOG) < 0){
        perror("otp_enc_d error listening on socket: ");
        exit(1);
    }
    //infinite loop to accept connections and pull from queued incoming connections
    while(1){
        //block accepting new connections until concurrent connections fall below 5
        while(child_pid_cnt == MAX_PIDS){
            CleanProcesses(&child_pid_cnt);
        }
        if(child_pid_cnt > 0){
            CleanProcesses(&child_pid_cnt);
        }
        sz_client_info = sizeof(client_addr);
        //accept connection on socket or block until one connects
        connection_fd = accept(socket_fd, (struct sockaddr*)&client_addr, &sz_client_info);
        if(connection_fd < 0){
            perror("otp_enc_d error accepting connection: ");
            continue;
        }
        else{
            //create child process
            child_pid = fork();
            switch(child_pid){
                case -1: //fork error
                    perror("otp_enc_d fork error: "); break;
                case 0: //child process
                    //recv message verifying id of otp_enc
                    if(VerifiedClient(connection_fd, "OTP_ENC")){
                        SendMsg(connection_fd, "success");
                    }
                    else{
                        close(connection_fd);
                        exit(2);
                        break;
                    }
                    //recv size of plaintext file from otp_enc
                    ptext_size = RecvSize(connection_fd);
                    if(ptext_size > 0){
                        //send message to otp_enc confirming size
                        SendSize(connection_fd, &ptext_size);
                    }
                    else{
                        fprintf(stderr, "File size of %d is invalid\n", ptext_size);
                        close(connection_fd);
                        exit(1);
                        break;
                    }
                    ptext = RecvTextfile(connection_fd, ptext_size);
                    if(ptext != NULL){
                        //printf("\n\n%s\n\n", ptext);
                        SendMsg(connection_fd, "success");
                    }
                    else{
                        fprintf(stderr, "File not received\n");
                        close(connection_fd);
                        exit(1);
                        break;
                    }
                    //recv size of keytext file from otp_enc
                    key_size = RecvSize(connection_fd);
                    if(key_size >= ptext_size){
                        //send message to otp_enc confirming size
                        SendSize(connection_fd, &key_size);
                    }
                    else{
                        fprintf(stderr, "Key size of %d is invalid\n", ptext_size);
                        close(connection_fd);
                        exit(1);
                        break;
                    }
                    ktext = RecvTextfile(connection_fd, key_size);
                    if(ptext != NULL){
                        //printf("\n\n%s\n\n", ktext);
                        SendMsg(connection_fd, "success");
                    }
                    else{
                        fprintf(stderr, "File not received\n");
                        close(connection_fd);
                        exit(1);
                        break;
                    }
                    //encode plaintext
                    //enc_text = EncodeText(ptext, ktext);
                    //send encoded text
                    free(ptext);
                    free(ktext);
                    close(connection_fd);
                    exit(0);
                    break;
                    //encode plaintext using key_file
                    //send ciphertext to otp_enc
                    //close connection socket
                default: //parent process
                    child_pid_cnt++;
            }
        }
    }
    return 0;
}
/**********************************************************************
 * altSIGTERM() - signal handler for SIGTERM
 **********************************************************************/
static void altSIGTERM(int signo){
    int status;
    //clean up child processes
    while(child_pid_cnt > 0){
        pid_t child_pid = waitpid(-1, &status, WNOHANG);
        if(child_pid > 0){
            child_pid_cnt--;
        }   
    }
    exit(0); //exit after clean up
}
/**********************************************************************
 * CleanProcesses() - checks for terminated child processes and cleans 
 * them up
 **********************************************************************/
void CleanProcesses(int* child_pid_cnt){
    int i, status;

    for(i = 0; i < *child_pid_cnt; i++){
        pid_t child_pid = waitpid(-1, &status, WNOHANG);
        if(child_pid > 0){
            (*child_pid_cnt)--;
        }
    }
}
/**********************************************************************
 * 
 **********************************************************************/

/***********************************************************************
char* EncodeText(char* plaintext, char* keytext){
    //enum char_set {A, 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 
    //              'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 
    //              'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', ' '};
    //enum char_set{A = 0, B = 1, C = 2, D, E, F, G, H, I,
    //              J, K, L, M, N, O, P, Q, R,
    //              S, T, U, V, X, Y, Z," " = 27};
    int i;
    enum char_set j;
    int* plaincode;
    int* keycode;
    int enc_length = strlen(plaintext);
    plaincode = calloc(enc_length, sizeof *plaincode);
    if(plaincode == NULL){
        perror("Error allocating memory for enc: ");
    }
    keycode = calloc(enc_length, sizeof *keycode);
    if(keycode == NULL){
        perror("Error allocating memory for enc: ");
    }

    for(i = 0; i < enc_length; i++){
        for(j = 'A'; j <= ' '; j++){
            if(plaintext[i] == j){
                plaincode[i] = j;
                break;
            }
        }
    }
    printf("%s[%d]\n", plaintext, enc_length);
    for(i = 0; i < enc_length; i++){
        printf("%d", plaincode[i]);
    }
    printf("[%d]\n", enc_length);
    
    free(plaincode);
    free(keycode);
    return NULL;
}
***********************************************************************/